Mathlib/Lean/Meta/RefinedDiscrTree/Initialize.lean

```diff
@@ -0,0 +1,298 @@
+  let { root, tries } := d
+  { root, tries := tries.map (.node #[] {} {}) }
+
+/-- Merge two discrimination trees. -/
+def append (x y : PreDiscrTree α) : PreDiscrTree α :=
+  let (x, y, f) :=
+        if x.root.size ≥ y.root.size then
+          (x, y, fun y x => x ++ y)
+        else
+          (y, x, fun x y => x ++ y)
+  let { root := yk, tries := ya } := y
+  yk.fold (init := x) fun d k yi => d.modifyAt k (f ya[yi]!)
+
+instance : Append (PreDiscrTree α) where
+  append := PreDiscrTree.append
+
+end PreDiscrTree
+
+
+/-- Information about a failed import. -/
+private structure ImportFailure where
+  /-- Module with constant that import failed on. -/
+  module  : Name
+  /-- Constant that import failed on. -/
+  const   : Name
+  /-- Exception that triggers error. -/
+  exception : Exception
+
+/-- Information generation from imported modules. -/
+private structure ImportErrorData where
+  errors : IO.Ref (Array ImportFailure)
+
+private def ImportErrorData.new : BaseIO ImportErrorData := do
+  return { errors := ← IO.mkRef #[] }
+/--
+Add the entries generated by `act name constInfo` to the `PreDiscrTree`.
+
+Note: It is expensive to create two new `IO.Ref`s for every `MetaM` operation,
+  so instead we reuse the same refs `mstate` and `cstate`. These are also used to
+  remember the cache, and the namegenerator across the operations.
+-/
+@[inline] private def addConstToPreDiscrTree
+    (cctx : Core.Context)
+    (env : Environment)
+    (modName : Name)
+    (data : ImportErrorData)
+    (mstate : IO.Ref Meta.State)
+    (cstate : IO.Ref Core.State)
+    (act : Name → ConstantInfo → MetaM (List (α × List (Key × LazyEntry))))
+    (tree : PreDiscrTree α) (name : Name) (constInfo : ConstantInfo) :
+    BaseIO (PreDiscrTree α) := do
+  -- here we use an if-then-else clause instead of the more stylish if-then-return,
+  -- because it compiles to more performant code
+  if constInfo.isUnsafe then pure tree else
+  if LazyDiscrTree.blacklistInsertion env name then pure tree else
+  /- For efficiency, we leave it up to the implementation of `act` to reset the states if needed -/
+  -- mstate.modify fun s => { cache := s.cache }
+  -- cstate.modify fun s => { env := s.env, cache := s.cache, ngen := s.ngen }
+  let mctx := { config := { transparency := .reducible } }
+  match ← (((act name constInfo) mctx mstate) cctx cstate).toBaseIO with
+  | .ok a =>
+    return a.foldl (fun t (val, entries) =>
+      entries.foldl (fun t (key, entry) => t.push key (entry, val)) t) tree
+  | .error e =>
+    let i : ImportFailure := {
+      module := modName,
+      const := name,
+      exception := e }
+    data.errors.modify (·.push i)
+    return tree
+
+
+/--
+Contains the pre discrimination tree and any errors occuring during initialization of
+the library search tree.
+-/
+private structure InitResults (α : Type) where
+  tree  : PreDiscrTree α := {}
+  errors : Array ImportFailure := #[]
```

Comment:
nit: the spacing here (and in other structures) is a little weird; I think the convention is to just use one space around `:`, but I guess column aligning is ok if you feel strongly. Right now, this is neither!